Dependencies/UrlLib/Source/Unix/UrlRequest.cpp

```diff
@@ -76,6 +75,80 @@ namespace UrlLib
         }
 
     private:
+        struct CurlMulti
+        {
+            CurlMulti()
+            : m_multiHandle(curl_multi_init())
+            {
+                assert(m_multiHandle != NULL);
+                curl_multi_setopt(m_multiHandle, CURLMOPT_MAX_TOTAL_CONNECTIONS, 200);
+                curl_multi_setopt(m_multiHandle, CURLMOPT_MAX_HOST_CONNECTIONS, 6L);
+                std::thread([this](){
+                    Loop();
+                }).detach();
+            }
+
+            ~CurlMulti()
+            {
+                auto errCode = curl_multi_cleanup(m_multiHandle);
+                assert(errCode == CURLM_OK);
+            }
+
+            void AddHandle(CURL* handle)
+            {
+                std::lock_guard<std::mutex> guard(m_mutex);
+                CURLMcode errCode = curl_multi_add_handle(m_multiHandle, handle); 
+                assert(errCode == CURLM_OK);
+                (void)errCode;
+            }
+
+        private:
+            CURLM* m_multiHandle;
+            // can't add new handle when curl is inside a callback. To be certain, a mutex is used when performing calls.
+            std::mutex m_mutex;
+
+            void Loop()
+            {
+                while (true) 
```

Comment:
You should probably use an arcana dispatcher for this instead of polling for things to do. Then you can use arcana to create tasks on the curl thread and pass data to it (like your question about task completion source). Maybe we can go over this offline.

Reply:
Curl multi interface handles everything once it's in its thread. The 'task' is added with ```curl_multi_add_handle```.